guardian/gateway
# Centralise res.locals manipulation.

## Description
Part II of the migration to an immutable object setup. This involves making a single place that `res.locals` mutation occurs, as well as potential state mutation which will be exempt from the mutation rules.

## Changes
* Create `requestStateMiddleware` middleware as a single place for generating `res.locals` based state.
* Create `fourZeroFourMiddleware` for handling 404 errors.
* Create `getABTesting` library function for setting up ABTesting.
* Rename type `ServerState` and `ResponseWithServerStateLocals` to `RequestState` and `ResponseWithRequestState` to better reflect the nature of that type (specific state per request).
* Rename `clientStateFromServerStateLocals()` to `clientStateFromRequestStateLocals()`
* Remove `applyRoutes`
* Remove `abTestsMiddleware`
* Remove `geoLocationMiddleware`
* Remove `mvtIdMiddleware`
* Remove `queryParamsMiddleware`
* Remove `serverStateLocalsMiddleware`
* Remove `updateCsrfPageUrlMiddleware`
* Remove `updateCsrfTokenMiddleware`
* `routeErrorHandler` no longer uses `res.locals`.
* Various documentation changes.

## Depends on
* #120
